File: data/src/commonMain/kotlin/com/voxenlabs/data/server/repositories/ServerRepository.kt
package com.voxenlabs.data.server.repositories

import com.voxenlabs.data.server.api.ServerApi
import com.voxenlabs.domain.server.models.Server
import com.voxenlabs.domain.server.models.ServerInfo
import com.voxenlabs.domain.server.repositories.ServerRepositoryInterface
import eu.anifantakis.lib.ksafe.KSafe
import org.koin.core.annotation.Singleton

@Singleton
class ServerRepository(
    val kSafe: KSafe,
    val serverApi: ServerApi,
) : ServerRepositoryInterface {
    override suspend fun getServerInfo(): ServerInfo = serverApi.getServerInfo().toDomain()

    override suspend fun getStoredServers(): List<Server> = kSafe.get(SERVER_KEY, defaultValue = listOf())

    override suspend fun storeServer(server: Server) {
        val servers = getStoredServers()

        kSafe.put(SERVER_KEY, servers + server)
    }

    override suspend fun removeStoredServer(url: String) {
        val servers = getStoredServers()

        kSafe.put(SERVER_KEY, servers.filterNot { it.url == url })
    }

    companion object {
        private const val SERVER_KEY = "servers"
    }
}

File: src/modules/Voxen.Client.Data/ServerDefinitions/Repositories/ServerRepository.cs
using Voxen.Client.Data.Storage;
using Voxen.Client.Domain.ServerDefinitions.Models;
using Voxen.Client.Domain.ServerDefinitions.RepositoryInterfaces;

namespace Voxen.Client.Data.ServerDefinitions.Repositories;

public class ServerRepository : IServerRepository
{
    private const string SERVER_KEY = "servers";

    private readonly IStorage<string> storage;

    public ServerRepository(IStorage<string> storage)
    {
        this.storage = storage;
    }

    /// <inheritdoc cref="IServerRepository"/>
    public List<Server> GetStoredServers()
    {
        return storage.Get<List<Server>?>(SERVER_KEY) ?? [];
    }

    /// <inheritdoc cref="IServerRepository"/>
    public void RemoveStoredServer(Server server)
    {
        var servers = GetStoredServers();
        servers.Remove(server);

        storage.Set(SERVER_KEY, servers);
    }

    /// <inheritdoc cref="IServerRepository"/>
    public void StoreServer(Server server)
    {
        var servers = GetStoredServers();
        servers.Add(server);

        storage.Set(SERVER_KEY, servers);
    }
}
